### PR TITLE
Propagate Errors from AsyncSource

### PR DESCRIPTION
**Pull request by oerling**
Adds error propagation from AsyncSource. If the function of the source
produces an error, the source is considered to have a value and
accessing the value with move() will throw the same error on every
call. move() still returns null if another thread is already waiting
for the source. The waiting thread is still sure to get the error.

### CODE REVIEW

File: velox/common/base/AsyncSource.h [hunk 1]
```diff
@@ -68,6 +73,12 @@ class AsyncSource {
     ContinueFuture wait;
     {
       std::lock_guard<std::mutex> l(mutex_);
+      // 'making_' can be read atomically, 'exception)' maybe
+      // not. Sotest 'making' so as not to read half-assigned
```
Comment (kgpai) [inline velox/common/base/AsyncSource.h:77]: nit: So test

File: velox/common/base/AsyncSource.h [hunk 2]
```diff
@@ -68,6 +73,12 @@ class AsyncSource {
     ContinueFuture wait;
     {
       std::lock_guard<std::mutex> l(mutex_);
+      // 'making_' can be read atomically, 'exception)' maybe
```
Comment (kgpai) [inline velox/common/base/AsyncSource.h:76]: nit : 'exception' not 'exception)'

File: velox/common/base/tests/AsyncSourceTest.cpp [hunk 3]
```diff
@@ -97,3 +103,46 @@ TEST(AsyncSourceTest, threads) {
...
+      } else {
+        // The rest of the threads get random gizmos. They are
+        // expected to produce an error or nullptr in the event
+        // another thread is already waiting for the same gizmo.
+        folly::Random::DefaultGenerator rng;
+        for (auto i = 0; i < kNumGizmos / 3; ++i) {
+          try {
+            auto gizmo =
+                gizmos[folly::Random::rand32(rng) % gizmos.size()]->move();
+            EXPECT_EQ(nullptr, gizmo);
```
Comment (kgpai) [inline velox/common/base/tests/AsyncSourceTest.cpp:137]: Lets add a counter to assert that we hit nullptrs ..

**Review by kgpai — APPROVED** [review]

File: velox/common/base/tests/AsyncSourceTest.cpp [hunk 4]
```diff
@@ -97,3 +103,50 @@ TEST(AsyncSourceTest, threads) {
...
+        }
+      }
+    }));
+  }
+  for (auto& thread : threads) {
+    thread.join();
+  }
+  // There will always be errors since the first to wait for any given
+  // gizmo is sure to get an error.
+  EXPECT_LT(0, numErrors);
```
Comment (xiaoxmeng) [inline velox/common/base/tests/AsyncSourceTest.cpp:151]: Shall we have a single thread and keep getting the error test cases? Thanks!

**Pull request closed without merging**